src/devices/cpu/upd177x/upd177x.cpp

```diff
@@ -0,0 +1,797 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+/*
+
+TODO:
+- Only upd1771c supported
+- Noise, time, and external interrupts not implemented
+- CH1/CH2 not implemented
+- Not all instructions implemented
+- Are PNC1 and PNC2 updated every machine cycle or only when the noise counter overflows?
+
+*/
+
+#include "emu.h"
+#include "upd177x.h"
+#include "upd177xd.h"
+
+#define VERBOSE 0
+#include "logmacro.h"
+
+
+DEFINE_DEVICE_TYPE(UPD1771C, upd1771c_cpu_device, "upd1771c", "NEC uPD1771C")
+
+
+enum
+{
+	UPD177X_PC, UPD177X_SP,
+	UPD177X_A, UPD177X_A_SAVE, UPD177X_H, UPD177X_N, UPD177X_X, UPD177X_Y,
+	UPD177X_TIMER, UPD177X_PNC1, UPD177X_PNC2,
+	UPD177X_PA, UPD177X_PB, UPD177X_MD0, UPD177X_MD1, UPD177X_SKIP, UPD177X_SKIP_SAVE
+};
+
+
+static constexpr u8 INT_TONE1 = 1 << 0;
+static constexpr u8 INT_TONE2 = 1 << 1;
+static constexpr u8 INT_TONE3 = 1 << 2;
+static constexpr u8 INT_TONE4 = 1 << 3;
+static constexpr u8 INT_NOISE = 1 << 4;
+static constexpr u8 INT_EXT = 1 << 5;
+static constexpr u8 INT_TIME = 1 << 6;
+
+static constexpr u8 MD0_3264_BIT = 0;
+static constexpr u8 MD0_TN_BIT = 1;
+static constexpr u8 MD0_NS_BIT = 2;
+static constexpr u8 MD0_NSF1_BIT = 3;
+static constexpr u8 MD0_NSF2_BIT = 4;
+static constexpr u8 MD0_OUT_BIT = 5;
+static constexpr u8 MD0_IF_BIT = 6;
+static constexpr u8 MD1_NSF3_BIT = 0;
+static constexpr u8 MD1_TIME_BIT = 1;
+static constexpr u8 MD1_EXT_BIT = 2;
+
+static constexpr u16 VECTOR_TONE1 = 0x20;
+static constexpr u16 VECTOR_TONE2 = 0x24;
+static constexpr u16 VECTOR_TONE3 = 0x28;
+static constexpr u16 VECTOR_TONE4 = 0x2c;
+static constexpr u16 VECTOR_NOISE = 0x48;
+static constexpr u16 VECTOR_EXT   = 0x60;
+static constexpr u16 VECTOR_TIME  = 0x80;
+
+static constexpr u8 STACK_START = 0x20;
+
+
+upd177x_cpu_device::upd177x_cpu_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock)
+	: cpu_device(mconfig, type, tag, owner, clock)
+	, device_sound_interface(mconfig, *this)
+	, m_program_config("program", ENDIANNESS_BIG, 16, 16, -1, address_map_constructor(FUNC(upd177x_cpu_device::program_map), this))
```

Comment:
Any hope that I can convince you to align with my MiSTer FPGA SCV core, and change the ROM to little-endian? Having two emulators which require different ROM endianness could cause chaos. :(

For my emulator, I chose endianness based on the behavior of the TBL* instructions. They byte-address the ROM in little-endian order, by selecting the lower byte when PR bit 0 is low (even). It also matches the RAM endianness for 16-bit stack accesses.

Reply:
Just to confirm: the file you use has a crc of 975bd68d?

Reply:
Answering my own question:
I see we would have the same checksums.
What an interesting coincidence, I was not aware that you were also working on a cpu core for the upd1771.

Reply:
I was also not aware of your work. No worries! It's a good time to be working on the Super Cassette Vision. The more people involved, the merrier. Happy to take what I've learned from the experience and help you.

Reply:
No worries, no one knew I was working on this; it started out as an experiment. I am normally quite bad at sound emulation so I am very happy that I got this far with it.
Your comments are very much appreciated. Thank you.